### PR TITLE
[Bugfix] Fix for sharding TP only

### PR DESCRIPTION
**Pull request by zarzen**
## Description ##
- When TP is enabled and PP is disabled, the last linear layer is sharded. Thus, we need `ParallelCrossEntropy` as the loss instead of the loss from HF. 
- Using TP with ZeRO-1 optimizer in DS runtime requires `mpu` as input to the `deespeed.initialize(...)` function. 
- When `--disable_pipeline` is used for disabling pipeline parallelism, we need to get the `num_mp` parameters from the `args`. 

## Known issue
- With the fix, the loss converge to about 7.x after 200 steps, which is still slower than ZeRO-1 or ZeRO-3 only. 
- For now only tested with GPT model. 

## Checklist ##

- [X] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [X] Code is well-documented

### CODE REVIEW

File: examples/utils.py [hunk 1]
```diff
@@ -99,18 +99,27 @@ def train_with_torch(
...
         loss = postproc(step, loss) if postproc is not None else loss
 
-        if step % 10 == 0:
-            logger.info(f"Step {step}, LOSS: {loss.item()}", ranks=0)
+        if model.global_steps % 10 == 0 and model.is_gradient_accumulation_boundary():
+            logger.info(f"Step {model.global_steps}, LOSS: {loss.item()}", ranks=0)
 
         loss = None
-        if step == steps:
+        if model.global_steps == steps:
```
Comment (comaniac) [inline examples/utils.py:124]: We could revert this change and rely on #31

File: examples/utils.py [hunk 2]
```diff
@@ -99,18 +99,27 @@ def train_with_torch(
...
         inputs, labels = preproc(step, batch) if preproc is not None else batch
-        loss = model(*inputs, labels=labels).loss
+        # print(f"inputs from the batch: {inputs}")
+        # print(f"labels from the batch: {labels}")
+        if hasattr(model, 'loss_fn'):
+            # use customized loss_fn due to sharded linear layer
+            outputs = model(*inputs)
+            loss = model.loss_fn(outputs.logits, labels)
+        else:
+            loss = model(*inputs, labels=labels).loss
```
Comment (comaniac) [inline examples/utils.py:109]: This might be confusing. Maybe we could do the following:

```python
def train_with_torch(
    model, dataloader, optimizer=None, loss_fn=None, ...
):
    ...
    if loss_fn is not None:
        outputs = model(*inputs)
        loss = model.loss_fn(outputs.logits, labels)
    else:
        loss = model(*inputs, labels=labels).loss
```
Actually I could also change this in #31 so that we don't need to resolve conflicts between PRs.

File: slapo/model_dialect/deepspeed/engine.py [hunk 3]
```diff
@@ -15,11 +17,17 @@ def init_ds_engine(model, **kwargs):
 
     if "config" not in kwargs:
         raise ValueError("DeepSpeed config not provided.")
+    mpu = None
+    if "topology" in kwargs:
+        mpu = kwargs["topology"]
+        if isinstance(mpu, PipeModelDataParallelTopology):
+            mpu = PipelineParallelGrid(topology=mpu)
```
Comment (comaniac) [inline slapo/model_dialect/deepspeed/engine.py:24]:
```suggestion
    mpu = kwargs.get("topology", None)
    if mpu is not None and isinstance(mpu, PipeModelDataParallelTopology):
        mpu = PipelineParallelGrid(topology=mpu)
```

File: examples/gpt/schedule.py [hunk 4]
```diff
@@ -5,8 +5,7 @@
 
 import torch
 import torch.nn as nn
-import torch.distributed as dist
-from torch.distributed.distributed_c10d import _get_global_rank
+import torch.distributed.distributed_c10d as dist
```
Comment (comaniac) [inline examples/gpt/schedule.py:8]: Is this the same as `torch.distributed`?
Also it is recommended to use `from torch import xxx as dist` instead.

File: examples/gpt/schedule.py [hunk 5]
```diff
@@ -416,11 +415,12 @@ def order_args_fn(*args, **kwargs):
 
 
 def broadcast_input(sch):
-    group_src_rank = _get_global_rank(sch.group, 0)
+    group_src_rank = dist.get_global_rank(sch.group, 0)
```
Comment (comaniac) [inline examples/gpt/schedule.py:418]: @szhengac please double check.

**Review by comaniac — APPROVED** [review]
Otherwise LGTM.

File: examples/gpt/schedule.py (hunk 5 above)
Comment (zarzen) [reply examples/gpt/schedule.py:418]: APIs start with underscore, is not encouraged to use. APIs in the `torch.distributed` package are routed to the `torch.distributed.distributed_c10d`

File: examples/gpt/schedule.py (hunk 4 above)
Comment (zarzen) [reply examples/gpt/schedule.py]: changed the import style

**Comment by comaniac** [comment]
Thanks @zarzen

**Comment by comaniac** [comment]
Found that this PR breaks 3D scripts (which are not covered in CI yet). @zarzen please file a fix.
